testplan/runnable/base.py

```diff
@@ -393,106 +394,130 @@ def schedule(self, task=None, resource=None, uid=None, **options):
         """
-        Adds a
-        :py:class:`runnable <testplan.common.entity.base.Runnable>`
-        tests entity to a :py:class:`~testplan.runners.base.Executor`
-        resource.
+        Adds a :py:class:`runnable <testplan.common.entity.base.Runnable>`
+        tests entity to a :py:class:`~testplan.runners.base.Executor` resource.
 
         :param runnable: Test runner entity.
         :type runnable: :py:class:`~testplan.common.entity.base.Runnable`
         :param resource: Test executor resource.
         :type resource: :py:class:`~testplan.runners.base.Executor`
-        :param uid: Optional test uid.
-        :type uid: ``str``
         :return: Assigned uid for test.
-        :rtype: ``str``
+        :rtype: ``str`` or ```NoneType``
         """
-        uid = uid or getattr(runnable, "uid", strings.uuid4)()
-        if uid in self._tests:
-            self.logger.error(
-                "Skip adding {} with uid {}.. already added.".format(
-                    runnable, uid
-                )
+        local_runner = self.resources.first()
+        resource = resource or local_runner
+
+        if resource not in self.resources:
+            raise RuntimeError(
+                'Resource "{}" does not exist.'.format(resource)
             )
-            return uid
 
-        if isinstance(runnable, Entity):
-            runnable.cfg.parent = self.cfg
+        # The runnable added into TestRunner can be: 1> a real test entity
+        # 2> a task wraps a test entity 3> a callble returns a test entity
+        if isinstance(runnable, Runnable):
             runnable.parent = self
+            runnable.cfg.parent = self.cfg
         elif isinstance(runnable, Task):
             pass
         elif callable(runnable):
-            runnable.parent_cfg = self.cfg
             runnable.parent = self
+            runnable.parent_cfg = self.cfg
 
-        # Check if test should not be added only when a filter is used.
-        if (
-            type(self.cfg.test_filter) is not filtering.Filter
-            or self.cfg.test_lister is not None
-        ):
-            if not self.should_be_added(runnable):
-                return None
+        # Get the real test target (a test entity)
+        target = self._verify_test_target(runnable)
+        if not target:
+            return None
 
-        local_runner = self.resources.first()
-        if resource is None:
-            resource = local_runner
+        uid = target.uid()
+        part = getattr(target.cfg, "part", None)
 
-        if resource not in self.resources:
-            raise RuntimeError(
-                'Resource "{}" does not exist.'.format(resource)
+        # When running interactively, add all real test targets into the local
+        # runner even if they were scheduled into a pool. It greatly simplifies
+        # the interactive runner if it only has to deal with the local runner.
+        if self.cfg.interactive_port is not None:
```

Comment:
Can we check interacitve mode in the beginning of add? If under interactive mode, just ignore the resource param, and all the rest logic shall be just the same? I want to avoid this if self.cfg.interactive_port as much as possible.

Reply:
ah. not very understand about avoid `self.cfg.interactive_port`